### ISSUE TITLE
torch.poisson returns floating point tensor

### ISSUE THREAD

**Issue by vishwakftw** [pytorch/pytorch#31779]
## 🐛 Bug

The Poisson distribution is a discrete distribution. As is the case with all other discrete distributions: `bernoulli` and `multinomial`, `torch.poisson` should also return a tensor with dtype `long` instead of `float`.

## To Reproduce

```python
import torch
torch.poisson(torch.rand(10))  # returns a float tensor
```

## Expected behavior

```
import torch
torch.poisson(torch.rand(10)) # returns a long tensor
```

**Comment by alicanb** [comment]
Sounds good, do you wanna take care of it?

**Comment by fritzo** [comment]
Sounds reasonable in principle, but changing `torch.distributions.Poisson` sample type will break lots of code, use 2x more memory, and be slower in computing `.log_prob()`. Could we add a `dtype` arg to `torch.poisson` and others so that, at least in the near term, `dist.Poisson` can still return `FloatTensor`s, or at least optionally return `FloatTensor`s?

Are there any plans to support `Int32` as NumPy does?

**Comment by vishwakftw** [comment]
We do support int32 as `torch.int`. I'm not sure about uint32.

Adding a dtype argument sounds like a good idea.

**Comment by alicanb** [comment]
@vishwakftw , I guess negative binomial will return `long` as well?

**Comment by vishwakftw** [comment]
@alicanb Yes, since Negative Binomial uses `torch.poisson`. If we choose to add a `dtype` arg, then this will change.

**Comment by neerajprad** [comment]
Even `torch.bernoulli` returns a float tensor (not long). IIRC, @apaszke mentioned that this was due to some legacy reasons and changing the dtype of these primitive functions would have led to a lot of code breakage. I think another reason for the current choice was that samples from `torch.multinomial` were often used to index into existing tensors, and it therefore made sense to have them be `LongTensor`, but samples from bernoulli and poisson were often multiplied with other float tensors so keeping their dtype as `float` made sense at the time. That said, I am not sure if these reasons are applicable any more since PyTorch will now implicitly cast to `float` without any issues, so returning a `LongTensor` should be backwards compatible.

@fritzo's suggestions of adding an optional dtype argument seems like a good idea which we can pipe in through distribution's `__init__`.

**Comment by ngimel** [comment]
Implicit casting comes with noticeable performance cost, so it still makes sense to return `float` where we expect a tensor to be used with other float tensors.

**Comment by fritzo** [comment]
> where we expect a tensor to be used with other float tensors.

E.g. in the context of probabilistic programming, Poisson samples are usually used later in `log_prob` computations.
